stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/MessageComposerController.kt

```diff
@@ -0,0 +1,246 @@
+package com.getstream.sdk.chat.viewmodel
+
+import com.getstream.sdk.chat.state.Edit
+import com.getstream.sdk.chat.state.MessageAction
+import com.getstream.sdk.chat.state.MessageMode
+import com.getstream.sdk.chat.state.Normal
+import com.getstream.sdk.chat.state.Reply
+import com.getstream.sdk.chat.state.Thread
+import com.getstream.sdk.chat.state.ThreadReply
+import io.getstream.chat.android.client.ChatClient
+import io.getstream.chat.android.client.models.Attachment
+import io.getstream.chat.android.client.models.Message
+import io.getstream.chat.android.offline.ChatDomain
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.map
+
+/**
+ * Controller responsible for handling the composing and sending of messages.
+ *
+ * It acts as a central place for both the core business logic and state required to create and send messages, handle
+ * attachments, message actions and more.
+ *
+ * If you require more state and business logic, compose this Controller with your code and apply the necessary changes.
+ *
+ * @param channelId The ID of the channel we're chatting in.
+ * @param chatClient The client used to communicate to the API.
+ * @param chatDomain The domain used to communicate to the API and store data offline.
+ */
+public class MessageComposerController constructor(
```

Comment:
This looks more like `MessageComposerViewModelDelegate` to me, but `MessageComposerController` is also fine 👍

Reply:
Isn't this supposed to be `@InternalStreamChatApi`? Do we ever want users to instantiate or call methods of this class?

Reply:
To call methods perhaps - they can reuse the core behavior if they want, without having to use the client directly I guess. 

They might also choose to not use ViewModels (as in Jetpack), so having a Controller they can wrap in their own presenter or Repo could be useful.

If you feel like it's not supposed to be exposed, we can always block it off!

Reply:
The fewer things we can expose as public API, the happier I am 😃 I'd like to keep this internal for the time being, and only expose it if we really, really need to. Integrating against the ViewModel should be enough for users. If they need low-level, they can use `ChatClient`. This feels like we're providing them too many options, so it'll just get complicated for them to choose between. Plus it's not consistent, we don't have such controllers available for other components.